### DISCUSSION TITLE
Enums not hydrating into instance, but staying as backed primitive

### DISCUSSION THREAD

**Discussion by x7ryan** [livewire/livewire#5041] [1. Help] [answered]
I am on Livewire 2.10.5

I setup an enum as a property like this:

````php
<?php

namespace App\Http\Livewire\TestComponent;

use App\Enums\Category;

class TestComponent extends Component
{
    public ?Category $category = null;
...
````

The Category enum is a string backed enum that looks like this:

````php
<?php

namespace App\Enums;

enum Category: string
{
    case CRYPTO = 'crypto';
    case ECOMMERCE = 'ecommerce';
    case FINTECH = 'fintech';
    case HEALTH = 'health';
    case LEGAL = 'legal';
    case OTHER = 'other';
}
````

Then I have a select element wire:modeled to the category property with the backed string as the value for each of the options. However attempting to update the property its not using the enum, its keeping the string version. I get this error: 

> Cannot assign string to property App\Http\Livewire\TestComponent::$category of type ?App\Enums\Category

If I remote the typehint or change it to `null|string|Category` It will save, but it will be as the string form. Livewire is not hydrating it to an instance of the Enum. 

It works in reverse. If I set a value, like adding this mount method:

````php
public function mount()
{
    $this->category = Category::OTHER;
}
````

Then it will dehydrate to the string value in the json, but it just never hydrates back to the actual Enum class in future requests.

Am I doing something wrong? Is this a bug?

**Comment by michael-rubel** [comment]
Hey, @x7ryan 

Only the objects listed in this section are properly supported by Livewire as public properties: https://laravel-livewire.com/docs/2.x/properties#important-notes
You can create private or protected ones of any type, but they would not be accessible from the frontend.

If you want to create a custom DTO/Enum available publicly, you should consider implementing `Wireable` property:
https://laravel-livewire.com/docs/2.x/properties#wireable-properties

**Reply by x7ryan** [reply]
Enums were actually added in February, [PR4506](https://github.com/livewire/livewire/pull/4506). 

But it seems per Josh Hanley's response below there is a limitation with the implementation.

**Comment by joshhanley** [comment, marked as answer]
@x7ryan Livewire supports hydrating and dehydrating enums when they are stored as public properties, but Livewire currently has no mechanism to handle an update of a property with an enum type, hence why it errors (same thing happens with carbon properties and others).

Currently we're working on a better and more configurable way to handle this.

But for now, to work around it, you need to add an updating handler into your component class (notice `$value` is passed by reference `&`)

```php
public function updatingCategory(&$value)
{
    $value = Category::from($value);
}
```

Hope this helps!

**Reply by x7ryan** [reply]
That's good to know. Certainly not what I would have expected. I would expect Livewire to call the from method for me, i mean so long as the the Enum class was type hinted so it could know what class to use.

Might be worth a quick note in the docs? I'll put together just a quick docs PR on this.

**Comment by jakewhiteley** [comment]
Are there any plans to add this natively now we are 1 year on?

**Reply by joshhanley** [reply]
@jakewhiteley this should already work in Livewire V3. Are you having issues with it?
